pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentSubscription.java

```diff
@@ -1201,6 +1203,21 @@ public long estimateBacklogSize() {
     }
 
     public SubscriptionStatsImpl getStats(GetStatsOptions getStatsOptions) {
+        // So far, there is no case hits this check.
+        if (getStatsOptions.isGetEarliestTimeInBacklog()) {
+            throw new IllegalArgumentException("Calling the sync method subscription.getStats with"
+                    + " getEarliestTimeInBacklog, it may encountered a deadlock error.");
+        }
+        // The method "getStatsAsync" will be a sync method if the param "isGetEarliestTimeInBacklog" is false.
+        try {
+            return getStatsAsync(getStatsOptions).get(5, TimeUnit.SECONDS);
```

Comment:
Maybe we should use `get()` instead of `get(5, TimeUnit.SECONDS)`?

Reply:
> Maybe we should use get() instead of get(5, TimeUnit.SECONDS)?

Both are OK, because it will be completed immediately. And `get(5, TimeUnit.SECONDS)` will release the deadlock after timeout, but `get()` will not

Reply:
You just reduce the lock time: 30s -> 5s, and we don't call this sync method in the pulsar, or maybe the sync method is only used for the unit test.

Reply:
> You just reduce the lock time: 30s -> 5s, and we don't call this sync method in the pulsar, or maybe the sync method is only used for the unit test.

Can not fully understand this comment

Reply:
The `metadata-store` thread will be released after 30 seconds, because the zk client will be reconnected(I remember).

So I say, your way just reduce the lock time to 5s.

The sync method has been deprecated, which is used for the unit test and removeSubscription.

<img width="908" alt="image" src="https://github.com/user-attachments/assets/faaeb106-205c-4b7d-9c21-976e670e9cfa">

Reply:
> The metadata-store thread will be released after 30 seconds, because the zk client will be reconnected(I remember).

It will not, metadata store thread is not the same with zk client internal thread

Reply:
> This thread will be released by org.apache.pulsar.metadata.impl.ZKSessionWatcher.

It is not related to the thread which working on `ping & pong`, and it will not trigger a timeout. You can do a test for it

Reply:
My point is that `get(5, TimeUnit.SECONDS)` doesn't make sense.

Reply:
> My point is that `get(5, TimeUnit.SECONDS)` doesn't make sense.

This is defensive programming, which can be removed, but it is better than removing it. After we modified the method `persistentTopic.removeSubscription`, we can remove the method `Subscription.getStat`